proxy/src/transport/tls/config.rs

```diff
@@ -110,27 +169,70 @@ impl CommonSettings {
     {
-        self.change_timestamps(interval)
-            .filter_map(move |_|
-                CommonConfig::load_from_disk(&self)
-                    .map_err(|e| warn!("error reloading TLS config: {:?}, falling back", e))
-                    .ok()
-            )
+        use inotify::{Inotify, WatchMask};
+        // If we used a less broad watch mask, we could avoid reloading the
+        // certs multiple times when k8s modifies a ConfigMap or Secret (a
+        // multi-step process that we see as a series of CREATE, MOVED_TO,
+        // MOVED_FROM, and DELETE events). However, this may not catch changes
+        // if the files we're watching *don't* live in a k8s ConfigMap/Secret,
+        // and I think false positives are much less harmful here than false
+        // negatives.
+        let mask = WatchMask::CREATE
+                 | WatchMask::MODIFY
+                 | WatchMask::DELETE
+                 | WatchMask::MOVE
+                 ;
+        let mut inotify = Inotify::init().map_err(Error::InotifyInit)?;
+
+        let paths = self.paths();
+        let paths = paths.into_iter()
+            .map(|path| {
+                // If the path to watch has a parent, watch that instead. This
+                // will allow us to pick up events to files in k8s ConfigMaps
+                // or Secrets (which we wouldn't detect if we watch the file
+                // itself, as they are double-symlinked).
+                //
+                // This may also result in some false positives (if a file we
+                // *don't* care about in the same dir changes, we'll still
+                // reload), but that's unlikely to be a problem.
+                let parent = path
+                    .parent()
+                    .map(Path::to_path_buf)
+                    .unwrap_or(path.to_path_buf());
+                trace!("will watch {:?} for {:?}", parent, path);
+                path
+            })
+            // Collect the paths into a `HashSet` eliminates any duplicates, to
+            // conserve the number of inotify watches we create.
+            //
+            // We could probably de-dup the iterator without allocating, by
+            // using `Iterator::peekable`, but this function isn't in the hot
+            // path, so this should be fine.
+            .collect::<HashSet<_>>();
+
+        for path in paths {
+            inotify.add_watch(path, mask)
+                .map_err(|e| Error::Io(path.to_path_buf(), e))?;
+            trace!("inotify: watch {:?}", path);
+        }
+
+        // Stream the events using `Inotify::into_event_stream` rather than
+        // `Inotify::event_stream`, so that we transfer ownership of the
+        // underlying file descriptor from the `Inotify` struct to the stream.
+        // Otherwise, the `Inotify` would close the fd when we drop it as we
+        // leave this function, rendering returned the stream useless.
```

Comment:
Again, I read this to describe an alternate method that we're not pursuing. It doesn't help me understand what's actually happening in the code that's implemented here.